### ISSUE TITLE
[Question] Use dashboard state to create custom events in "Scheduler events" widget

### ISSUE THREAD

**Issue by DurandA** [thingsboard/thingsboard#9999]
**Component**
 * UI

**Description**
I am using ThingsBoard PE and is interested in customizing the "Scheduler events" widget. Specifically, I am unable to configure/implement the following features using the default widget when creating a new event:
1. select the originator based on a selected entity from the dashboard state
2. customize the form input of a custom event type based on an attribute of a selected entity from the dashboard state
 
Given these limitations and since the "Scheduler events" widget is not open-source, I believe the second best option is to create a "Custom action (with HTML template)" to create a new event so I can implement my own logic. I was however unable to create a `SCHEDULER_EVENT` entity using the public APIs. Here is what we tried:
 
```javascript
let $injector = self.ctx.$scope.$injector;
let entityService = $injector.get(self.ctx.servicesMap.get('entityService'));
let event = {
      "originatorId": null,
      "name": "foo",
      "type": "demoEvent",
      "schedule": {
            "timezone": "Europe/Zurich",
            "startTime": 1705014000000,
            "repeat": {
                  "type": "DAILY",
                  "endsOn": 1705446000000
            }
      },
      "configuration": {
            "msgBody": {
                  "myDemoField": "baz"
            },
            "metadata": {
                  "qux": "quxx"
            },
            "originatorId": null,
            "msgType": null
      }
};
entityService.saveEntityParameters("SCHEDULER_EVENT", event, false);
```
 
This unfortunately throws the following error `ERROR TypeError: this.getSaveEntityObservable(...) is undefined`.

**Environment**

 * ThingsBoard: 3.6.1PE

**Comment by devaskim** [comment]
You need the following API
```
const event = {
...
};
const scheduler = self.ctx.$scope.$injector.get(self.ctx.servicesMap.get('schedulerEventService'));
scheduler.saveSchedulerEvent(event);
```

**Comment by DurandA** [comment]
@devaskim Thank you, that is very helpful.

Using the `schedulerEventService` works as expected. Since it is not part of the source code of this repository/the CE version, is this API documented somewhere?

**Comment by devaskim** [comment]
No, I guess. Only reverse engineering with the help of `console.log(...)`

**Comment by DurandA** [comment]
Thanks for the answer.

**Comment by DurandA** [comment]
I have one more question concerning the "Scheduler events" widget with custom actions.

Once I created or updated a scheduler event from a custom action, how can I update the widget data (so the list of scheduler events is refreshed)? With typical widgets containing normal entities, I would call the following after saving the entity:
```javascript
widgetContext.updateAliases();
vm.dialogRef.close(null);
```
However, it seems that the scheduler events do not use a datasource based on an alias.

Upon inspection of the widgetContext, I tried calling various functions without success so far:
```javascript
widgetContext.reset();
widgetContext.detectChanges();
widgetContext.detectContainerChanges();
widgetContext.updateWidgetParams();
```

**Comment by devaskim** [comment]
Again, you have magic `console.log()` for reverse engineering . 

1. So, go to the Widget Library -> open Scheduled Events widget.
2. There you have access to the component `tb-scheduler-events` with the help of `self.ctx.$scope.schedulerEventsWidget`. 
3. So add `console.log(self.ctx.$scope.schedulerEventsWidget)` on `self.onInit` callback and click Run.
4. Then open browser DevTools and look at Console tab.
5. Find the output of `console.log` and explore widget object.

https://github.com/thingsboard/thingsboard/assets/86909129/c35c2d37-f983-48e8-bc5e-2cd3879a9f8c

I guess, after adding an event you need to call `self.ctx.$scope.schedulerEventsWidge.updateData()`

**Comment by DurandA** [comment]
Inspecting objects was what I did, but I somehow completely missed that there were a different widget reference in the scope.

Thank you for your detailed answer, this is very much appreciated.
